aten/src/ATen/native/LossNLL.cpp

```diff
@@ -82,43 +95,75 @@ static void nll_loss_out_frame(
   const scalar_t* input_data = input_contiguous.data_ptr<scalar_t>();
   const int64_t* target_data = target_contiguous.data_ptr<int64_t>();
 
-  scalar_t output_val = 0;
-  scalar_t total_weight_val = 0;
+  const int64_t ndim = input.dim();
+  TORCH_CHECK(ndim == 1 || ndim == 2);
+  const int64_t batch_size = ndim == 1 ? 1 : input.size(0);
+  TORCH_CHECK(target.size(0) == batch_size);
+  const int64_t map_size = n_classes;
+  const int64_t numiter = batch_size * map_size;
+
+  constexpr int64_t cascade_sum_num_levels = 8;
+  scalar_t weight_partial_sums[cascade_sum_num_levels] = {0};
+  scalar_t loss_partial_sums[cascade_sum_num_levels] = {0};
+  const int64_t level_power =
+      std::max(int64_t(4), ceil_log2(numiter) / cascade_sum_num_levels);
+  const int64_t level_step = (1 << level_power);
+  const int64_t level_mask = level_step - 1;
+
+  for (int64_t b = 0; b < batch_size; b++) {
+    for (int64_t elem = 0; elem < map_size; elem++) {
```

Comment:
why is there a loop over map_size (which is n_classes)? Will it lead to same data summed multiple times in lines 129-135?
Previously there was a loop over batch size only.

Reply:
Fixed and test are all passing again. Thanks!